### ISSUE TITLE
[no-unused-vars] False positives with optional chaining

### ISSUE THREAD

**Issue by tryggvigy** [typescript-eslint/typescript-eslint#2875]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
Just upgraded our very large repo to
```diff
-    "@typescript-eslint/eslint-plugin": "3.9.1",
-    "@typescript-eslint/parser": "3.9.1",
+    "@typescript-eslint/eslint-plugin": "4.9.1",
+    "@typescript-eslint/parser": "4.9.1",
-    "typescript": "3.9.6",
+    "typescript": "4.1.3",
```

and seeing a lot of `@typescript-eslint/no-unused-vars` errors. They seem to be false positives that trigger either by [optional arguments](https://github.com/typescript-eslint/typescript-eslint/issues/2874#issuecomment-744419401) or optional chaining. I have a minimal case here for optional chaining:

![foo](https://user-images.githubusercontent.com/2373958/102084235-c4f74080-3e0c-11eb-80ca-d5032f7d9442.gif)

As you can see removing optional chaining fixes the issue.

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "error",
  }
}
```

```TS
function toURI(a: any): any {
  return a;
}

export const isSameContext = (
  oneContextUri?: string | null,
  anotherContextUri?: string | null,
): boolean => {
  if (!oneContextUri || !anotherContextUri) return false;

  const contextUri1 = toURI(oneContextUri);
  const contextUri2 = toURI(anotherContextUri);

  return contextUri1?.isSameIdentity(contextUri2);
};
```

**Expected Result**
No error from `@typescript-eslint/no-unused-vars` when using optional chaining

**Actual Result**
`@typescript-eslint/no-unused-vars` reports errors when using optional chaining

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.1` |
| `@typescript-eslint/parser`        | `4.9.1` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.14.0` |
| `node`                             | `14.15.1` |

**Comment by bradzacher** [comment]
As mentioned in https://github.com/typescript-eslint/typescript-eslint/issues/2874 - I cannot repro this against master.

Also as mentioned in that issue - You're likely running into this due to your installed version of the parser being wrong.

In v4 we breaking changed the AST representation for optional chaining, which is not compatible with ESLint's representation.

**Comment by tryggvigy** [comment]
Sorry for the false positive. `rm -rf node_modules/` did the trick
